### PR TITLE
feat(cloudflare): use unenv implementation of `setImmediate` for `node:timers`

### PR DESCRIPTION
**Pull request by vicb**
We are exploring using unenv for `setImmediate`.
The desired behavior is run the callback in a `setTimeout(0)` instead of right away.

### CODE REVIEW

File: src/runtime/node/timers/internal/immediate.ts [hunk 1]
```diff
@@ -17,3 +17,31 @@ export class Immediate<TArgs extends any[]> implements NodeJS.Immediate {
...
+
+export class ImmediateTimeoutBased<TArgs extends any[]>
+  implements NodeJS.Immediate
+{
+  _onImmediate: (...args: TArgs) => void;
+  private _timeout: NodeJS.Timeout;
+
+  constructor(callback: (...args: TArgs) => void, args: TArgs) {
+    this._onImmediate = callback;
+    this._timeout = setTimeout(callback, 0, ...args);
```
Comment (pi0) [inline src/runtime/node/timers/internal/immediate.ts:29]: I think you can directly update `Immediate` class implementation to make use of timeout (and we can extract this change as a separate PR)

Comment (vicb) [reply src/runtime/node/timers/internal/immediate.ts:29]: So the change would not be cf specific?

Comment (pi0) [reply src/runtime/node/timers/internal/immediate.ts:29]: Yes! Using setTimeout seems obviously better choice than immediately calling fn

**Comment by vicb** [comment]
@IgorMinar @pi0 helped me to get https://github.com/unjs/unenv/pull/303 in - it changes the `setImmediate` behavior to use a timeout.

The current behavior is to use `globalThis || unenv`.

This PR would make sure we always use `unenv`.

File: src/runtime/node/timers/$cloudflare.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  unenroll,
+} from "./index";
+
+import {
+  setImmediateFallback as setImmediate,
+  clearImmediateFallback as clearImmediate,
+} from "./internal/set-immediate";
+
+globalThis.setImmediate = setImmediate;
+globalThis.clearImmediate = clearImmediate;
```
Comment (IgorMinar) [inline src/runtime/node/timers/$cloudflare.ts:38]: the challenge with this implementation is that this module is now side-effecty which is what we've been trying to avoid in all modules in  unenv except for the `inject` config here: https://github.com/unjs/unenv/blob/f1c0bb8f0feeb22537e15b7f443af352dfb94777/src/presets/cloudflare.ts#L62-L69

could you please remove these two lines and add two new entries to `inject` section of the cloudflare.ts preset?

Comment (IgorMinar) [reply src/runtime/node/timers/$cloudflare.ts:38]: you can use `globalThis.console` or `process` as implemented today via `inject` as the pattern to follow. thank you

**Review by IgorMinar — CHANGES_REQUESTED** [review]

**Comment by vicb** [comment]
So...

```js
  inject: {
    setImmediate: ["unenv/runtime/node/timers/$cloudflare", "setImmediate"],
    clearImmediate: ["unenv/runtime/node/timers/$cloudflare", "clearImmediate"],
  },
```

was not working with this code before the PR

```js
// set-immediate.ts

export const clearImmediateFallback = function clearImmediate(...){...};
```

Because the code generated by wrangler was

```js
// At this point clearImmediateFallback === undefined
globalThis.clearImmediate = clearImmediateFallback;

// ...

var setImmediateFallback = function (...) {...};
var clearImmediateFallback = function (...) {...};

// setImmediate is ok
globalThis.setImmediate = setImmediatefFallback;
```

By switching exports to functions, it solves the issue, the generated code is now

```js
// At this point clearImmediateFallback === undefined
globalThis.clearImmediate = clearImmediateFallback;

// ...

function setImmediateFallback(...) {...}
function clearImmediateFallback(...) {...}

// setImmediate is ok
globalThis.setImmediate = setImmediatefFallback;

```

Note: wrangler is using an older version of `esbuild` but updating to the current `0.23.1` still exhibits the same behavior.

**Review by IgorMinar — APPROVED** [review]
bah! I'm sorry @vicb, this must have been frustrating to debug. I'm glad that you got to the bottom of it and resolved it. Thank you!

The rest looks great 🧡

File: src/runtime/node/timers/$cloudflare.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+export const setTimeout: typeof timers.setTimeout =
+  globalThis.setTimeout || setTimeoutFallback;
+export const setInterval: typeof timers.setInterval =
+  globalThis.setInterval || setIntervalFallback;
+
+export const active = notImplemented("timers.active");
+export const _unrefActive = notImplemented("timers._unrefActive");
+export const enroll = notImplemented("timers.enroll");
+export const unenroll = notImplemented("timers.unenroll");
```
Comment (pi0) [inline src/runtime/node/timers/$cloudflare.ts:37]: For exports that are identical to unenv built-in, I think we can simply reexport.

File: src/runtime/node/timers/internal/set-immediate.ts
Comment (pi0) [inline src/runtime/node/timers/internal/set-immediate.ts:1]: That's fine for PR but please try to reduce diff in PRs if it is only code style change.

**Comment by vicb** [comment]
Thanks for the review and help @IgorMinar and @pi0.

I have updated the PR according to the feddback.

**Comment by vicb** [comment]
> bah! I'm sorry @vicb, this must have been frustrating to debug. I'm glad that you got to the bottom of it and resolved it. Thank you!

It was fun to scrath our heads with @petebacondarwin this morning 🤔

File: src/runtime/node/timers/internal/set-interval.ts [hunk 4]
```diff
@@ -1,18 +1,17 @@
...
+  value?: T,
+): AsyncIterable<T> {
+  yield value as T;
+}
 
 export function setIntervalFallback(
   callback: (args: void) => void,
   ms?: number,
 ): NodeJS.Timeout;
+
```
Comment (vicb) [inline src/runtime/node/timers/internal/set-interval.ts:14]: actually this should be better reverted. Let me force push one last time

**Review by pi0 — APPROVED** [review]
